### ISSUE TITLE
MinimalAPI configuration reload do not  update the the options value

### ISSUE THREAD

**Issue by WeihanLi** [dotnet/aspnetcore#34056]
### Describe the bug

When I use the MinimalAPI, I met an issue that the value of the options could not get updated when I try to reload the configuration via `IConfigurationRoot.Reload()`, I'm not sure if this is by design.

Sample code is as below:

``` c#
var builder = WebApplication.CreateBuilder(args);
builder.Services.Configure<SettingsOption>(builder.Configuration);
var app = builder.Build();

var logger = app.Services.GetRequiredService<ILoggerFactory>()
    .CreateLogger("Test");
app.Map(new PathString("/redeploy"), appBuilder => appBuilder.Run(context =>
{
    var config = context.RequestServices.GetRequiredService<IConfiguration>();
    logger.LogInformation($"Configuration in redeploy hash code: {config.GetHashCode()}");
    if (config is IConfigurationRoot configuration)
    {
        var currentSlot = configuration["Setting1"];
        configuration["Setting1"] = "Slot2" != currentSlot ? "Slot2" : "Slot1";
        configuration.Reload();
        return context.Response.WriteAsync("Success");
    }
    logger.LogWarning($"configuration is not configuration root, configurationType: {config.GetType().FullName}");
    return Task.CompletedTask;
}));
app.MapFallback(context =>
{
    var option = context.RequestServices.GetRequiredService<IOptionsMonitor<SettingsOption>>();
    return context.Response.WriteAsJsonAsync(new{ option.CurrentValue.Setting1, Time= DateTime.Now });
});
app.Run();
```

When I access the `redeploy` endpoint, the value of the option should be updated I think.

**More**

After some research, I found that the `builder.Configuration` and `IConfiguration` from DI are not the same instance, so when I call `reload`, the option's value did not update, maybe this is by design? I'm not sure so I created this issue.

### To Reproduce

There's a simple sample for reproducing the issue, <https://github.com/WeihanLi/AspNetCorePlayground/tree/master/MinimalAPIConfigurationIssue>

1. Run `dotnet run`
2. Access the endpoint <http://localhost:5000>
3. Access the endpoint <http://localhost:5000/redeploy>
4. Reply step 2, access <http:localhost:5000> again

Expected behavior:

The `setting1` value from the response should be `Slot2`

Actucal behavior:

The `setting1` value from the response keep as `null`

### Further technical details
- ASP.NET Core version: .NET 6 Preview 5
- Include the output of `dotnet --info`
![image](https://user-images.githubusercontent.com/7604648/124343461-2c84aa00-dbfe-11eb-8678-fc7acceb13ea.png)

**Comment by halter73** [comment]
This might be fixed by https://github.com/dotnet/aspnetcore/issues/33082

**Comment by WeihanLi** [comment]
Thanks @halter73, maybe I should close this issue now or wait for the fix?

**Comment by halter73** [comment]
Let's keep it open for now @WeihanLi. This provides a good test case, so thanks! I close it once I confirmed it's fixed.

**Comment by halter73** [comment]
Thanks again for the report @WeihanLi. It turns out this was fixed by #33081 in preview6 which is now available.

![image](https://user-images.githubusercontent.com/54385/125862451-041aad75-2dba-4f9e-a96c-1677fe00d62b.png)

I feel it's kind of cheating because we simply register `WebApplicationBuilder.Configuration` as an `IConfiguration` service after the inner `HostBuilder`'s `IConfiguration`.


https://github.com/dotnet/aspnetcore/blob/632b24193d12991c7d4844728a8fd3846a6aefc3/src/DefaultBuilder/src/WebApplicationBuilder.cs#L47-L48

This works when the `IConfiguration` is resolved from DI by middleware as is done in this sample, but there are some places where the inner `HostBuilder` is still [using it's own _hostConfiguration](https://github.com/dotnet/runtime/blob/dfdec3bfbb09a2d9d865651925b1679dd1f5aa7a/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L259), so it would be best to also use a ChainedConfigurationSource as suggested by #33082. I'll continue tracking that with this issue.

**Comment by WeihanLi** [comment]
> it would be best to also use a ChainedConfigurationSource as suggested by #33082

Agree with this, thanks @halter73
